File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        vector<ListNode*> v;
        while(head){v.push_back(head);head=head->next;}
        for(int i=0;i<v.size()/k;i++){
            for(int j=0;j<k/2;j++){
                swap(v[k*i+j],v[k*i+k-1-j]);
            }
        }
        for(int i=0;i<v.size()-1;i++){
            v[i]->next=v[i+1];
        }v.back()->next=NULL;
        return v[0];
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head, ListNode end){
        
        ListNode prev = null;
        
        while(head!=end){
            ListNode nextnode = head.next;
            head.next = prev;
            prev = head;
            head= nextnode;
            
        }
        return prev;
    }
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head==null){
            return null;
        }
        ListNode curr=head;
        ListNode tail = curr;
        for (int i=0;i<k; i++){
            if (tail==null){
                return head;
            }
            tail = tail.next;
        }
        ListNode newhead = reverse(head, tail);
        head.next = reverseKGroup(tail, k);
        return newhead;

    }
}